### ISSUE TITLE
Confusion over block changes

### ISSUE THREAD

**Issue by telegraphic** [ledatelescope/bifrost#73]
I'm having trouble porting my code to make it work with the most recent changes. Could I please get some help bringing this example code back up?

```python
import numpy as np 
import time
import bifrost.pipeline as bfp
from bifrost.fft import fft
import bifrost as bf
from copy import deepcopy

np.set_printoptions(precision=2)


# Runtime options

n_acc = 5
n_iter = 5


# Generate test vector
dx = 1
dy = 1
dz = 100

TEST_VEC = np.zeros((dx, dy, dz))
for hh in range(dx):
    for ii in range(dy):
        TEST_VEC[hh, ii] = np.arange(dz)
TEST_VEC = np.expand_dims(TEST_VEC, axis=0)

def timeit(method):
    """ Decorator for timing execution of a method in a class """
    def timed(*args, **kw):
        ts = time.time()
        result = method(*args, **kw)
        te = time.time()
        
        classname  = repr(args[0]).split(' ')[0][10:]
        methodname = str(method.__name__)
        time_str   = '%2.2fs' % (te-ts)
        
        print '%24s %16s %16s' % (classname, methodname, time_str)
        return result
    return timed


class FakeReader(object): 
    """ Simple file-like reading object for pipeline testing
    
    Args:
        count_max (int): number of times to generate data 
        test_vec (np array): test vector to be returned
    """
    def __init__(self, filename, test_vec, count_max):
        super(FakeReader, self).__init__()
        self.count = 0
        self.count_max = count_max
        self.data = test_vec
        
    def read(self):
        self.count += 1
            
        if self.count <= self.count_max:
            return self.data
        else:
            self.close()
            return np.array([])
        
    def __enter__(self):
        return self
    
    def close(self):
        pass
    
    def __exit__(self, type, value, tb):
        self.close()



class FakeReadBlock(bfp.SourceBlock):
    """ Block for generating dummy data 
    
    Args:
        count_max (int): number of times to generate data 
        test_vec (np array): test vector to be placed into output ring
    """
    def __init__(self, sourcenames, gulp_nframe, test_vec, count_max, *args, **kwargs):
        super(FakeReadBlock, self).__init__(sourcenames, gulp_nframe, *args, **kwargs)
        self.test_vec = test_vec
        self.count_max = count_max
        
        
    def create_reader(self, sourcename):
        print "Loading %s" % sourcename
        return FakeReader(sourcename, self.test_vec, self.count_max)
         
    def on_sequence(self, ireader, sourcename):        
        ohdr = {'name': sourcename,
                '_tensor': {
                        'dtype':  'cf32',
                        'shape':  [-1, dx, dy, dz],
                        }, 
                }
        return [ohdr]
    
    @timeit
    def on_data(self, reader, ospans):
        
        ospan  = ospans[0]
        indata = reader.read()
        nframe = indata.shape[0]
        
        #print ospan.data.shape, indata.shape, nframe
        if ospan.data[:nframe].shape == indata.shape:
            ospan.data[0] = indata
            #assert np.allclose(ospan.data[0], indata)
        return [nframe]


class GpuFFTBlock(bfp.TransformBlock):
    """ FFT block, takes FFT of intput data """
    def __init__(self, iring,  *args, **kwargs):
    	super(GpuFFTBlock, self).__init__(iring, *args, **kwargs)

    def on_sequence(self, iseqs):
        ohdr = deepcopy(iseqs.header)
        return [ohdr]
    
    def on_data(self, ispans, ospans):
        fft(ispans.data.as_BFarray(), ospans.data.as_BFarray())
        return [1]


class CheckBlock(bfp.SinkBlock):
    """ Block for validating output data """
    def __init__(self, iring, *args, **kwargs):
        super(CheckBlock, self).__init__(iring, *args, **kwargs)
        
    def on_sequence(self, iseqs):
        ohdr = deepcopy(iseqs.header)
        return [ohdr]
        pass
        
    def on_data(self, ispans):
        print(ispans.data.shape)
        print(ispans.data[0])
        #assert np.allclose(ispans[0].data, np.fft.fft(TEST_VEC))

class TestPipeline(object):
    
    @timeit
    def run(self):
       filenames   = ['one.file']
       b_read      = FakeReadBlock(filenames, 1, test_vec=TEST_VEC, count_max=10, core=0)
       b_fft       = GpuFFTBlock(b_read.orings[0], core=1)
       b_print     = CheckBlock(b_fft.orings[0], core=2)
           
       pipeline = bfp.get_default_pipeline()
       
       print(pipeline.dot_graph())
    
       pipeline.run() 

if __name__ == "__main__":

    test = TestPipeline()
    test.run()

```

Which fails with:

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 222, in run
    self.main(active_orings)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 324, in main
    ohdr['time_tag'] = self._seq_count
TypeError: list indices must be integers, not str
```

Essentially I'm confused over what should be a list of spans, a list of headers, or just a span or a header. If I change the TransformBlock `on_sequence` to `return ohdr` instead of `return [ohdr]` I get:

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 222, in run
    self.main(active_orings)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 336, in main
    islices = [_span_slice(slice_) for slice_ in islices]
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 304, in _span_slice
    start = soft_slice.start or 0
AttributeError: 'dict' object has no attribute 'start'
```

and if I change the SourceBlock to return `ohdr` instead of `[ohdr]` I get 

```
Exception in thread FakeReadBlock_0:  (<--- I changed my pipeline.py to get this)
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 222, in run
    self.main(active_orings)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 269, in main
    ohdr['time_tag'] = self._seq_count
TypeError: 'str' object does not support item assignment
```

**Comment by benbarsdell** [comment]
Sorry about this, I ran into the same trouble when porting existing blocks.

Basically, where there was a list before, now there is just a single thing, and SinkBlock has no output.
Here is a version that runs:

```
import numpy as np 
import time
import bifrost.pipeline as bfp
from bifrost.fft import fft
import bifrost as bf
from copy import deepcopy

np.set_printoptions(precision=2)


# Runtime options

n_acc = 5
n_iter = 5


# Generate test vector
dx = 1
dy = 1
dz = 100

TEST_VEC = np.zeros((dx, dy, dz))
for hh in range(dx):
    for ii in range(dy):
        TEST_VEC[hh, ii] = np.arange(dz)
TEST_VEC = np.expand_dims(TEST_VEC, axis=0)

def timeit(method):
    """ Decorator for timing execution of a method in a class """
    def timed(*args, **kw):
        ts = time.time()
        result = method(*args, **kw)
        te = time.time()
        
        classname  = repr(args[0]).split(' ')[0][10:]
        methodname = str(method.__name__)
        time_str   = '%2.2fs' % (te-ts)
        
        print '%24s %16s %16s' % (classname, methodname, time_str)
        return result
    return timed


class FakeReader(object): 
    """ Simple file-like reading object for pipeline testing
    
    Args:
        count_max (int): number of times to generate data 
        test_vec (np array): test vector to be returned
    """
    def __init__(self, filename, test_vec, count_max):
        super(FakeReader, self).__init__()
        self.count = 0
        self.count_max = count_max
        self.data = test_vec
        
    def read(self):
        self.count += 1
            
        if self.count <= self.count_max:
            return self.data
        else:
            self.close()
            return np.array([])
        
    def __enter__(self):
        return self
    
    def close(self):
        pass
    
    def __exit__(self, type, value, tb):
        self.close()



class FakeReadBlock(bfp.SourceBlock):
    """ Block for generating dummy data 
    
    Args:
        count_max (int): number of times to generate data 
        test_vec (np array): test vector to be placed into output ring
    """
    def __init__(self, sourcenames, gulp_nframe, test_vec, count_max, *args, **kwargs):
        super(FakeReadBlock, self).__init__(sourcenames, gulp_nframe, *args, **kwargs)
        self.test_vec = test_vec
        self.count_max = count_max
        
        
    def create_reader(self, sourcename):
        print "Loading %s" % sourcename
        return FakeReader(sourcename, self.test_vec, self.count_max)
         
    def on_sequence(self, ireader, sourcename):        
        ohdr = {'name': sourcename,
                '_tensor': {
                        'dtype':  'cf32',
                        'shape':  [-1, dx, dy, dz],
                        }, 
                }
        return [ohdr]
    
    @timeit
    def on_data(self, reader, ospans):
        
        ospan  = ospans[0]
        indata = reader.read()
        nframe = indata.shape[0]
        
        #print ospan.data.shape, indata.shape, nframe
        if ospan.data[:nframe].shape == indata.shape:
            ospan.data[0] = indata
            #assert np.allclose(ospan.data[0], indata)
        return [nframe]


class GpuFFTBlock(bfp.TransformBlock):
    """ FFT block, takes FFT of intput data """
    def __init__(self, iring,  *args, **kwargs):
    	super(GpuFFTBlock, self).__init__(iring, *args, **kwargs)

    def on_sequence(self, iseq):
        ohdr = deepcopy(iseq.header)
        return ohdr
    
    def on_data(self, ispan, ospan):
        fft(ispan.data.as_BFarray(), ospan.data.as_BFarray())
        return 1


class CheckBlock(bfp.SinkBlock):
    """ Block for validating output data """
    def __init__(self, iring, *args, **kwargs):
        super(CheckBlock, self).__init__(iring, *args, **kwargs)
        
    def on_sequence(self, iseq):
        pass
        
    def on_data(self, ispan):
        print(ispan.data.shape)
        print(ispan.data[0])
        #assert np.allclose(ispans[0].data, np.fft.fft(TEST_VEC))

class TestPipeline(object):
    
    @timeit
    def run(self):
       filenames   = ['one.file']
       b_read      = FakeReadBlock(filenames, 1, test_vec=TEST_VEC, count_max=10, core=0)
       b_fft       = GpuFFTBlock(b_read.orings[0], core=1)
       b_print     = CheckBlock(b_fft.orings[0], core=2)
           
       pipeline = bfp.get_default_pipeline()
       
       print(pipeline.dot_graph())
    
       pipeline.run() 

if __name__ == "__main__":

    test = TestPipeline()
    test.run()
```

**Comment by telegraphic** [comment]
Thanks, so the source block still outputs the header in a list?

So this code runs, but the output is all zeros. For a reason I'm yet to track down, the fft is now throwing an error code when I try and run the above. It also looks like the `_check` function is expecting arguments?

Here's the block with the _check function, in case I'm using it wrong:

```python
class GpuFFTBlock(bfp.TransformBlock):
    """ FFT block, takes FFT of intput data """
    def __init__(self, iring,  *args, **kwargs):
    	super(GpuFFTBlock, self).__init__(iring, *args, **kwargs)

    def on_sequence(self, iseq):
        ohdr = deepcopy(iseq.header)
        return ohdr
    
    def on_data(self, ispan, ospan):
        print type(ospan.data), type(ispan.data)
        _check(fft(ispan.data.as_BFarray(), ospan.data.as_BFarray()))
        return 1
```

And the output

```
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 222, in run
    self.main(active_orings)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 371, in main
    ostrides = self._on_data(ispans, ospans)
  File "build/bdist.linux-x86_64/egg/bifrost/pipeline.py", line 433, in _on_data
    nframe_commit = self.on_data(ispans[0], ospans[0])
  File "bf_fft_block.py", line 130, in on_data
    _check(fft(ispan.data.as_BFarray(), ospan.data.as_BFarray()))
  File "build/bdist.linux-x86_64/egg/bifrost/libbifrost.py", line 118, in _check
    status, args = f
TypeError: 'int' object is not iterable
```

**Comment by benbarsdell** [comment]
Yes sorry I have yet to update SourceBlock; I meant to mention it before.

`_check` is an internal function; you shouldn't need to use it, because fft should be using it internally. Except that fft is not using it, which is a bug:

https://github.com/ledatelescope/bifrost/blob/master/python/bifrost/fft.py#L41
should actually be `_check(_bf.FFT(...`

**Comment by MilesCranmer** [comment]
@benbarsdell that's my fault, sorry. I will fix it now.

**Comment by MilesCranmer** [comment]
Fixed with 776639bceb23e5d18f11bb72c5a31eb860dd73dd

**Comment by telegraphic** [comment]
closing this as it's been addressed.
